src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -47,131 +37,56 @@
+    public AppiumCommandExecutor(Map<String, CommandInfo> additionalCommands,
+                                 URL addressOfRemoteServer) {
         this(additionalCommands, addressOfRemoteServer, new ApacheHttpClient.Factory());
     }
 
-    public AppiumCommandExecutor(Map<String, AppiumCommandInfo> additionalCommands,
-        DriverService service) {
+    public AppiumCommandExecutor(Map<String, CommandInfo> additionalCommands,
+                                 DriverService service) {
         this(additionalCommands, service, new ApacheHttpClient.Factory());
     }
 
-    public URL getAddressOfRemoteServer() {
-        return remoteServer;
-    }
-
-    private Response doExecute(Command command) throws IOException, WebDriverException {
-        if (command.getSessionId() == null) {
-            if (QUIT.equals(command.getName())) {
-                return new Response();
-            }
-            if (!GET_ALL_SESSIONS.equals(command.getName())
-                    && !NEW_SESSION.equals(command.getName())) {
-                throw new NoSuchSessionException(
-                        "Session ID is null. Using WebDriver after calling quit()?");
-            }
-        }
-
-        if (NEW_SESSION.equals(command.getName())) {
-            if (commandCodec != null) {
-                throw new SessionNotCreatedException("Session already exists");
-            }
-            AppiumProtocolHandShake handshake = new AppiumProtocolHandShake();
-            AppiumProtocolHandShake.Result result = handshake.createSession(client, command);
-            Dialect dialect = result.getDialect();
-            commandCodec = dialect.getCommandCodec();
-
-            additionalCommands.forEach((key, value) -> {
-                checkNotNull(key);
-                checkNotNull(value);
-                commandCodec.defineCommand(key, value.getMethod(), value.getUrl());
-            } );
-
-            responseCodec = dialect.getResponseCodec();
-            return result.createResponse();
-        }
-
-        if (commandCodec == null || responseCodec == null) {
-            throw new WebDriverException(
-                    "No command or response codec has been defined. Unable to proceed");
-        }
-
-        HttpRequest httpRequest = commandCodec.encode(command);
-        try {
-            HttpResponse httpResponse = client.execute(httpRequest, true);
-
-            Response response = responseCodec.decode(httpResponse);
-            if (response.getSessionId() == null) {
-                if (httpResponse.getTargetHost() != null) {
-                    response.setSessionId(HttpSessionId.getSessionId(httpResponse.getTargetHost()));
-                } else {
-                    response.setSessionId(command.getSessionId().toString());
+    @Override public Response execute(Command command) throws WebDriverException {
+        if (DriverCommand.NEW_SESSION.equals(command.getName())) {
+            ofNullable(service).ifPresent(driverService -> {
```

Comment:
I assume the code will be simpler if we make the service variable itself of type Optional